Add `GenericSequence`, `Lengthen`, `Shorten`, `Split` and `Concat` traits.

novacrazy opened the pull request:
Here are some more useful features that allow for easier passing of `GenericArray` via generics, as well as some simple resize methods implemented as inverse traits. I've had to use these a few times already.

These really should have been in that last pull request, but I wasn't far enough along in a dependent project to really test it.

novacrazy commented:
The whole "inverse trait" idea is useful because of arrays like `GenericArray<T, U0>`, where it can't be shortened anymore but could be lengthened. However, the lengthened array should be able to be shortened again to get the original `GenericArray<T, U0>`. 

In `type Longer: Shorten<T, Shorter=Self>` (and `Shorten` trait equivalent), the `Shorter=Self` clause allows Rust to correctly deduce that type of `self.lengthen(_).shorten().0` is equal to `Self`, which is awesome.

novacrazy commented:
In fact, would you like the idea of me adding `head`, `tail`, `init` and `last` methods for segmenting arrays on the type level? I could probably implement a generic `split` method and use it for the others.

fizyk20 commented:
Wow, very nice ideas! Yes, the methods you mention sound awesome, it would be great if you added them :)

novacrazy commented:
A few changes which cover your other pull request.

I've redesigned the `Lengthen` and `Shorten` traits. They are still inverse traits, but instead of `lengthen` and `shorten` methods, we now have `append`, `prepend`, `pop_back` and `pop_front`. These are more familiar and follow more traditional terminology. 

I've also added the `Split` and `Concat` traits. These are NOT inverse traits for now. I'm not sure they even could be, given how arbitrary they are. 

Also, I've simplified the logic I use for these. Since it's literally just moving/copying data, I've followed your example and simplified things to raw pointer operations. They can't panic, so it's fine.

Overall, I'd say it's pretty complete. If you're happy with this then I can't think of anything else to hold off for.

fizyk20 commented:
Thanks! And sorry for the delay in merging :)

novacrazy commented:
So you're going to hate me for this, but I promise I'm not trying to make things harder on you.

I'm working on some new improvements to this stuff that should be ready in a week or two.

I just wanted to warn you before doing that semver trick thing in case you are willing to wait.
